opm/simulators/wells/StandardWell_impl.hpp

```diff
@@ -573,16 +575,17 @@ namespace Opm
                 EvalWell cq_r_thermal(this->numWellEq_ + Indices::numEq, 0.);
                 if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx) && FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
 
-                    if(FluidSystem::waterPhaseIdx == phaseIdx)
+                    // d = 1.0 - rs * rv
```

Comment:
I am not talking about the change in this PR, to me, the for loop `for (unsigned phaseIdx = 0; phaseIdx < FluidSystem::numPhases; ++phaseIdx) ` is a little problematic. Will get back to this very soon.

Reply:
The things looking strange here is that, `if(FluidSystem::waterPhaseIdx == phaseIdx)`, we do not need to handle the things related to `rs`, `rv` and `d` at all. 

As a result, should not the `if` structure should be something like the following, 
```c++
            for (unsigned phaseIdx = 0; phaseIdx < FluidSystem::numPhases; ++phaseIdx) {
                if (!FluidSystem::phaseIsActive(phaseIdx)) {
                    continue;
                }

                EvalWell cq_r_thermal(this->numWellEq_ + Indices::numEq, 0.);
                const unsigned activeCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::solventComponentIndex(phaseIdx));
                const bool both_oil_gas = FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx) && FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx);
                if(FluidSystem::waterPhaseIdx == phaseIdx || ~both_oil_gas) {
                    cq_r_thermal = cq_s[activeCompIdx] / this->extendEval(fs.invB(phaseIdx));
                } else {
                    // remove dissolved gas and vapporized oil
                    const unsigned oilCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::oilCompIdx);
                    const unsigned gasCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
                    // q_os = q_or * b_o + rv * q_gr * b_g
                    // q_gs = q_gr * g_g + rs * q_or * b_o
                    // d = 1.0 - rs * rv
                    const EvalWell d = this->extendEval(1.0 - fs.Rv() * fs.Rs());
                    // q_gr = 1 / (b_g * d) * (q_gs - rs * q_os)
                    if(FluidSystem::gasPhaseIdx == phaseIdx)
                        cq_r_thermal = (cq_s[gasCompIdx] - this->extendEval(fs.Rs()) * cq_s[oilCompIdx]) / (d * this->extendEval(fs.invB(phaseIdx)) );
                    // q_or = 1 / (b_o * d) * (q_os - rv * q_gs)
                    if(FluidSystem::oilPhaseIdx == phaseIdx)
                        cq_r_thermal = (cq_s[oilCompIdx] - this->extendEval(fs.Rv()) * cq_s[gasCompIdx]) / (d * this->extendEval(fs.invB(phaseIdx)) );

                }
```

Or I miss something here? 

Sorry, I am talking about the master branch, not the change in the PR at the moment. However, the problem I think in the master branch is shown by the content in this PR `                    if(d <= 0 || FluidSystem::waterPhaseIdx == phaseIdx)` here.

Reply:
I don't think the code your propose is any different than what is in the PR with the exception of how many times you compute `const EvalWell d = this->extendEval(1.0 - fs.Rv() * fs.Rs());` or did I misunderstand your comment. This calculations could be moved outside the loop over the phases if performance is an issue here.

Reply:
In the original code, 
```c++
                if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx) && FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {

                    if(FluidSystem::waterPhaseIdx == phaseIdx)
                        cq_r_thermal = cq_s[activeCompIdx] / this->extendEval(fs.invB(phaseIdx));

                    // remove dissolved gas and vapporized oil
                    const unsigned oilCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::oilCompIdx);
                    const unsigned gasCompIdx = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
                    // q_os = q_or * b_o + rv * q_gr * b_g
                    // q_gs = q_gr * g_g + rs * q_or * b_o
                    // d = 1.0 - rs * rv
                    const EvalWell d = this->extendEval(1.0 - fs.Rv() * fs.Rs());
                    // q_gr = 1 / (b_g * d) * (q_gs - rs * q_os)
                    if(FluidSystem::gasPhaseIdx == phaseIdx)
                        cq_r_thermal = (cq_s[gasCompIdx] - this->extendEval(fs.Rs()) * cq_s[oilCompIdx]) / (d * this->extendEval(fs.invB(phaseIdx)) );
                    // q_or = 1 / (b_o * d) * (q_os - rv * q_gs)
                    if(FluidSystem::oilPhaseIdx == phaseIdx)
                        cq_r_thermal = (cq_s[oilCompIdx] - this->extendEval(fs.Rv()) * cq_s[gasCompIdx]) / (d * this->extendEval(fs.invB(phaseIdx)) );

                } else {
                    cq_r_thermal = cq_s[activeCompIdx] / this->extendEval(fs.invB(phaseIdx));
                }
```

Assuming black oil case, there are oil, gas and water. 

`if(FluidSystem::waterPhaseIdx == phaseIdx)`, 

after
```c++
                    if(FluidSystem::waterPhaseIdx == phaseIdx)
                        cq_r_thermal = cq_s[activeCompIdx] / this->extendEval(fs.invB(phaseIdx));
```

We still need do calculate `rs`, `rv`, `d`, and they will NOT be applied to `cq_r_thermal`, because there other two if guards there, 
```
                    if(FluidSystem::gasPhaseIdx == phaseIdx)
                        cq_r_thermal = (cq_s[gasCompIdx] - this->extendEval(fs.Rs()) * cq_s[oilCompIdx]) / (d * this->extendEval(fs.invB(phaseIdx)) );
                    // q_or = 1 / (b_o * d) * (q_os - rv * q_gs)
                    if(FluidSystem::oilPhaseIdx == phaseIdx)
                        cq_r_thermal = (cq_s[oilCompIdx] - this->extendEval(fs.Rv()) * cq_s[gasCompIdx]) / (d * this->extendEval(fs.invB(phaseIdx)) );
```

Reply:
With 
```c++
                } else {
                    cq_r_thermal = cq_s[activeCompIdx] / this->extendEval(fs.invB(phaseIdx));
                }
```
The result is probably okay, while it is confusing `if` structure and there are non-necessary calculation of `rs`, rv`, and `d`.

Reply:
I agree with you that this part of the code could be cleaned up a bit. Your suggestion looks good, but as you said this clean-up is unrelated to the content of this PR, and should be done in a following up PR IMO.

Reply:
Yes. But it removes the necessity of the following change in this PR, because they do not need to be together.  I do not insist, a cleaning up PR can be okay. 

```
if(d <= 0.0 || FluidSystem::waterPhaseIdx == phaseIdx) {
```

Reply:
I created a cleanup PR based on the above discussion, 

https://github.com/OPM/opm-simulators/pull/3855

Maybe it is good to get that one first, so we can review the change in this part of the code?